lib/core/src/Cardano/Pool/DB/Sqlite.hs

```diff
@@ -196,62 +200,114 @@ newDBLayer trace fp = do
-                    let poolMargin = unsafeMkPercentage $ toRational $ marginNum % marginDen
+                    let poolMargin = unsafeMkPercentage $
+                            toRational $ marginNum % marginDen
                     let poolCost = Quantity poolCost_
                     let poolPledge = Quantity poolPledge_
-                    let poolMetadata = (,) <$> poolMetadataUrl <*> poolMetadataHash
-                    poolOwners <- fmap (poolOwnerOwner . entityVal) <$> selectList
-                        [ PoolOwnerPoolId ==. poolId ]
-                        [ Desc PoolOwnerSlot, Asc PoolOwnerIndex ]
-                    pure $ Just $ PoolRegistrationCertificate
-                        { poolId
-                        , poolOwners
-                        , poolMargin
-                        , poolCost
-                        , poolPledge
-                        , poolMetadata
-                        }
+                    let poolMetadata =
+                          (,) <$> poolMetadataUrl <*> poolMetadataHash
+                    poolOwners <- fmap (poolOwnerOwner . entityVal) <$>
+                        selectList
+                            [ PoolOwnerPoolId
+                                ==. poolId
+                            , PoolOwnerSlot
+                                ==. slotId
+                            , PoolOwnerSlotInternalIndex
+                                ==. slotInternalIndex
+                            ]
+                            [ Asc PoolOwnerIndex ]
+                    let cert = PoolRegistrationCertificate
+                            { poolId
+                            , poolOwners
+                            , poolMargin
+                            , poolCost
+                            , poolPledge
+                            , poolMetadata
+                            }
+                    let cpt = CertificatePublicationTime
+                            {slotId, slotInternalIndex}
+                    pure $ Just (cpt, cert)
+
+        , putPoolRetirement = \cpt cert -> do
+            let CertificatePublicationTime {slotId, slotInternalIndex} = cpt
+            let PoolRetirementCertificate
+                    { poolId
+                    , retiredIn
+                    } = cert
+            let EpochNo retirementEpoch = retiredIn
+            repsert (PoolRetirementKey poolId slotId slotInternalIndex) $
+                PoolRetirement
+                    poolId
+                    slotId
+                    slotInternalIndex
+                    (fromIntegral retirementEpoch)
+
+        , readPoolRetirement = \poolId -> do
+            let filterBy = [ PoolRetirementPoolId ==. poolId ]
+            let orderBy =
+                    [ Desc PoolRetirementSlot
+                    , Desc PoolRetirementSlotInternalIndex
+                    ]
+            selectFirst filterBy orderBy >>= \case
+                Nothing -> pure Nothing
+                Just meta -> do
```

Comment:
How about this way?

```suggestion
            res <- selectFirst [ PoolRetirementPoolId ==. poolId ]
                    [ Desc PoolRetirementSlot
                    , Desc PoolRetirementSlotInternalIndex
                    ]
            forM res $ \meta -> do
```

Reply:
> How about this way?

Nice!

The function `readPoolRegistration` was originally written this way, so I extended the style when writing `readPoolRetirement`. But your way is shorter and clearer.

I've applied it to the pair of functions `readPool{Registration,Retirement}`

Fixed in: c79a3300883dfc4bd62c18ec23aa730bb4a208b6.